dbms/src/Common/BackgroundSchedulePool.cpp

```diff
@@ -0,0 +1,259 @@
+    : size(size)
+{
+    LOG_INFO(&Logger::get("BackgroundSchedulePool"), "Create BackgroundSchedulePool with " << size << " threads");
+
+    threads.resize(size);
+    for (auto & thread : threads)
+        thread = std::thread([this] { threadFunction(); });
+
+    delayed_thread = std::thread([this] { delayExecutionThreadFunction(); });
+}
+
+
+BackgroundSchedulePool::~BackgroundSchedulePool()
+{
+    try
+    {
+        shutdown = true;
+        wakeup_event.notify_all();
+        queue.wakeUpAll();
+
+        delayed_thread.join();
+
+        LOG_TRACE(&Logger::get("BackgroundSchedulePool"), "Waiting for threads to finish.");
+        for (std::thread & thread : threads)
+            thread.join();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+
+BackgroundSchedulePool::TaskHandle BackgroundSchedulePool::addTask(const std::string & name, const Task & task)
+{
+    return std::make_shared<TaskInfo>(*this, name, task);
+}
+
+
+void BackgroundSchedulePool::removeTask(const TaskHandle & task)
+{
+    task->deactivate();
+}
+
+
+void BackgroundSchedulePool::scheduleDelayedTask(const TaskHandle & task, size_t ms, std::lock_guard<std::mutex>&)
+{
+    Poco::Timestamp current_time;
+
+    {
+        std::lock_guard lock(delayed_tasks_lock);
+
+        if (task->delayed)
+            delayed_tasks.erase(task->iterator);
+
+        task->iterator = delayed_tasks.emplace(current_time + (ms * 1000), task);
+        task->delayed = true;
+    }
+
+    wakeup_event.notify_all();
+}
+
+
+void BackgroundSchedulePool::cancelDelayedTask(const TaskHandle & task, std::lock_guard<std::mutex> &)
+{
+    {
+        std::lock_guard lock(delayed_tasks_lock);
+        delayed_tasks.erase(task->iterator);
+        task->delayed = false;
+    }
+
+    wakeup_event.notify_all();
+}
+
+
+void BackgroundSchedulePool::threadFunction()
+{
+    setThreadName("BackgrSchedPool");
+
+    MemoryTracker memory_tracker;
+    memory_tracker.setMetric(CurrentMetrics::MemoryTrackingInBackgroundSchedulePool);
+    current_memory_tracker = &memory_tracker;
+
+    while (!shutdown)
+    {
+        if (Poco::AutoPtr<Poco::Notification> notification = queue.waitDequeueNotification())
+        {
+            TaskNotification & task_notification = static_cast<TaskNotification &>(*notification);
+            task_notification.execute();
+        }
+    }
+
+    current_memory_tracker = nullptr;
+}
+
+
+void BackgroundSchedulePool::delayExecutionThreadFunction()
+{
+    setThreadName("BckSchPoolDelay");
+
+    while (!shutdown)
+    {
+        TaskHandle task;
+
+        {
+            std::unique_lock lock(delayed_tasks_lock);
+            Poco::Timestamp min_time;
+
+            if (!delayed_tasks.empty())
+            {
+                auto t = delayed_tasks.begin();
+                min_time = t->first;
+                task = t->second;
+            }
+
+            if (!task)
+            {
+                wakeup_event.wait(lock);
```

Comment:
There is a small possibility of deadlock on shutdown here. If this thread checks the shutdown variable and then sleeps for some reason before waiting on the condition, it can miss the notification from destructor (line 142) and wait forever.

The wakeup_event field is misnamed - it is a std::condition_variable, not Poco::Event (which has different semantics).

Reply:
We can fix the deadlock by adding  std::lock_guard lock(delayed_tasks_lock); in desturctor before shutdown = true. What do you think ?

Reply:
That would still leave a tiny window for deadlock - because shutdown flag is not checked under the lock in the delayExecutionThreadFunction loop.

Reply:
Right we can add a check inside the lock as well on the first line.

Reply:
Fixed in commit 6629b03af9bb4ebf50b2ab94b7c3a965b773907c